### PR TITLE
misc: added benches

### PR DESCRIPTION
**Pull request by YuJuncen**
This PR added some benches to the repo.
Sadly, most of benches shows we should improve the performance in the future...

### CODE REVIEW

File: benches/spawn_depth.rs [hunk 1]
```diff
@@ -0,0 +1,53 @@
+use criterion::{criterion_group, criterion_main, Criterion};
```
Comment (HuSharp) [inline benches/spawn_depth.rs]: add Copyright?

File: README.md [hunk 2]
```diff
@@ -94,7 +94,12 @@ tracing-active-tree = { git = "https://github.com/tikv/tracing-active-tree.git",
 
 ## Benchmark
 
-TODO
+There are overheads when tracing. Generally, tracing wide(a root future waits for many futures) trees has poorer performance
```
Comment (pingyu) [inline README.md]: The new line is intended ?

```suggestion
There are overheads when tracing. Generally, tracing wide(a root future waits for many futures) trees has poorer performance than tracing deep(a tree like a linked list) tree.
```

File: README.md [hunk 3]
```diff
@@ -94,7 +94,12 @@ tracing-active-tree = { git = "https://github.com/tikv/tracing-active-tree.git",
...
+than tracing deep(a tree like a linked list) tree.
+
```
Comment (pingyu) [inline README.md]: Suggest to give some result of benchmarks. I think users would be concerned about "how poor" it would be. So do I :)

Comment (YuJuncen) [reply README.md]: I'm not sure... Each time I have different result when I'm running the bench mark.

Comment (pingyu) [reply README.md]: OK, forget it... But let's discuss later on how to quantify the overhead brought by the tracing & the trees.

**Review by HuSharp — APPROVED** [review]
LGTM!

File: README.md [hunk 4]
```diff
@@ -94,7 +94,37 @@ tracing-active-tree = { git = "https://github.com/tikv/tracing-active-tree.git",
 
 ## Benchmark
 
-TODO
+There are overheads when tracing. When tracing huge task trees, the overhead may obvious. Run benches in the `benches` folder to check it.
```
Comment (pingyu) [inline README.md]:
```suggestion
There are overheads when tracing. When tracing huge task trees, the overhead may be obvious. Run benches in the `benches` folder to check it.
```

File: README.md [hunk 5]
```diff
@@ -94,7 +94,37 @@ tracing-active-tree = { git = "https://github.com/tikv/tracing-active-tree.git",
...
+
+You may also enable the feature `coarsetime` for a better performance of fetching system time.
+
+There are some examples of the overheading, generally, huge (both wide or deep) trees will bring performance penalty:
```
Comment (pingyu) [inline README.md]:
```suggestion
Here are some examples of the overheads. Generally, huge (both wide or deep) trees will bring performance penalty:
```

File: README.md [hunk 6]
```diff
@@ -94,7 +94,37 @@ tracing-active-tree = { git = "https://github.com/tikv/tracing-active-tree.git",
...
+
+```console
+> cargo bench --features="coarsetime" --bench deep_tree
+deep_baseline           time:   [194.64 µs 208.66 µs 223.05 µs]
+deep_with_subs          time:   [4.7118 ms 4.8690 ms 5.0579 ms]
+```
+
+Tracing wide(a root future waits for many futures) trees has poorer performance than tracing deep(a tree like a linked list) tree:
```
Comment (pingyu) [inline README.md]:
```suggestion
Tracing wide(a root future waits for many futures) trees have poorer performance than tracing deep(a tree like a linked list) trees:
```

File: README.md [hunk 7]
```diff
@@ -94,7 +94,37 @@ tracing-active-tree = { git = "https://github.com/tikv/tracing-active-tree.git",
...
+
+```console
+> cargo bench --features="coarsetime" --bench wide_tree
+wide_baseline           time:   [895.63 µs 924.51 µs 953.05 µs]
+wide_with_subs          time:   [16.192 ms 17.383 ms 18.629 ms]
+```
+
+> **NOTE**
+> 
+> When you tweaking the parameters of those benchmarks, you may notice that the performance regression isn't linear with the amount of tree nodes. This imples that it is possible to optimize this. They should be linear.
```
Comment (pingyu) [inline README.md]:
```suggestion
> When you tweaking the parameters of those benchmarks, you may notice that the performance regression isn't linear with the amount of tree nodes. This implies that it is possible to optimize this. They should be linear.
```

File: README.md [hunk 8]
```diff
@@ -94,7 +94,37 @@ tracing-active-tree = { git = "https://github.com/tikv/tracing-active-tree.git",
...
+
+But there isn't observable overheading when spawning many tiny tasks:
```
Comment (pingyu) [inline README.md]:
```suggestion
But there isn't observable overheads when spawning many tiny tasks: 
```

File: benches/many_tasks.rs [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+#[instrument]
+async fn dummy(n: usize) {
+    for _ in 0..10 {
+        tokio::task::yield_now().await;
+    }
+}
+
+pub fn spawn_many_tasks(n: usize) {
+    let futs = (0..n).map(|i| tokio::spawn(dummy(i)));
```
Comment (pingyu) [inline benches/many_tasks.rs:17]: Need a `root!` or not ?

File: benches/many_tasks.rs [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+use criterion::{criterion_group, criterion_main, Criterion};
+
+use tracing::instrument;
+use tracing_active_tree::layer;
+use tracing_subscriber::prelude::*;
+
+#[instrument]
+async fn dummy(n: usize) {
+    for _ in 0..10 {
+        tokio::task::yield_now().await;
```
Comment (pingyu) [inline benches/many_tasks.rs:12]: Need a `frame!` ?

File: benches/many_tasks.rs (hunk 9 above)
Comment (YuJuncen) [reply benches/many_tasks.rs:17]: I guess we don't need a root for instrumented async calls.

File: benches/many_tasks.rs (hunk 10 above)
Comment (YuJuncen) [reply benches/many_tasks.rs:12]: We can but I think it isn't necessary. The main goal of this case is for testing many small tasks being spawned.🤔

**Review by pingyu — APPROVED** [review]
LGTM~

**Pull request merged**